### PR TITLE
Add multi-lock support for atomic multi-lock.

### PR DESCRIPTION
**Pull request by bhallionOhbibi**
Main changes
===
The purpose of thoses changes is to be able to obtain multiple lock in a single atomic operation.
This is useful in many situation were you need to lock resources access by multiple process.

For exemple, if we have:
 - Process A accessing resource `1`
 - Process B accessing resources `1` & `2`

We want Process A to be able to lock resource 1 and therefore preventing Process B to access it until Process A is done.

I'v been siting on this update since August 2022 and have been using it on production without issues.

Other changes
===
- Update test to use unique lockKeys each run.

### CODE REVIEW

**Comment by bhallionOhbibi** [comment]
Is there any chance someone takes a look a this ? :smile:

**Comment by dim** [comment]
@bhallionOhbibi apologies for the delay, but we have been exceptionally busy lately. We will take a look in the next few days

File: redislock.go
Comment (mxmCherry) [inline redislock.go:97]: Auto-generating prefix - I don't like that.

Let user decide. Just use what user provides.

Doesn't prevent us from merging and then polishing it ourselves before tagging a release.

File: redislock.go
Comment (mxmCherry) [inline redislock.go:128]: Ideally, prealloc this - `setArgs := make([]string, 0, len(keys) * 2)`.

Or, maybe better call the msetnx script with more usual KEYS=the_keys_array, ARGV=[value, ttl]. Then build the msetnx args in script to save some "traffic".

Doesn't prevent us from merging and then polishing it ourselves.

File: redislock.go
Comment (mxmCherry) [inline redislock.go:171]: Emm... I doubt it saves a lot of code, I'd better remove.

Doesn't prevent us from merging and then polishing it ourselves.

File: redislock.go
Comment (mxmCherry) [inline redislock.go:173]: Comment doesn't match the method name.

Doesn't prevent us from merging and then polishing it ourselves.

File: redislock.go
Comment (mxmCherry) [inline redislock.go:136]: As I mentioned in msetnx script above - pretty much a no-op.

The script could just return boolean on its own. No need to expose count of set-expires to the caller.

Doesn't prevent us from merging and then polishing it ourselves.

File: redislock.go
Comment (mxmCherry) [inline redislock.go:211]: Same "script could return boolean" note here, I guess.

Doesn't prevent us from merging and then polishing it ourselves.

File: redislock.go
Comment (mxmCherry) [inline redislock.go:231]: One more case when script could return true/false on its own.

Doesn't prevent us from merging and then polishing it ourselves.

File: redislock_test.go
Comment (mxmCherry) [inline redislock_test.go:258]: Again - the lib should better not assume any prefixes on its own - just use what user provides.

Doesn't prevent us from merging and then polishing it ourselves.

File: redislock_test.go
Comment (mxmCherry) [inline redislock_test.go:398]: Nit: weird line skip.

Doesn't prevent us from merging and then polishing it ourselves.

File: release.lua
Comment (mxmCherry) [inline release.lua:10]: I'd prefer the script to return bool.

Doesn't prevent us from merging and then polishing it ourselves.

File: msetnxex.lua
Comment (mxmCherry) [reply msetnxex.lua:6]: UPD: remembered the case when key could hyphotetically be gone - if it would expire while this loop is running.

But found this: https://stackoverflow.com/questions/49158805/is-this-redis-lua-script-that-deals-with-key-expire-race-conditions-a-pure-funct

Basically, ppl say that key can expire, but won't be removed while script runs.

So my initial `error(...)` instead of softer expire-success counting is still valid.

File: msetnxex.lua
Comment (mxmCherry) [inline msetnxex.lua:1]: Nit: here and for all other scripts. I'd maybe group/embed them from smth like `./scripts/XXX.lua`.

Matter of taste, doesn't prevent us from merging.

**Review by mxmCherry — COMMENTED** [review]
@dim from my PoV - this doesn't contain any obvious bugs.

I only spotted some places that raised questions, but we can address them ourselves.

I'll do another pass on tests tomorrow or next week - pretty much ignored them for now.

But then - I think we can merge this and polish to our taste / optimizations on our own before tagging next release.

**Comment by mxmCherry** [comment]
Note: I'm all in for the Lua scripts instead of using Redis Pipelines here. We don't need any retvals from loop-ed batch operations (done within loops here as there's no / I have found no multi-key-EXPIRE command etc), so let's not waste "traffic" if we can do all the checks we want right in Lua.

File: redislock_test.go
Comment (mxmCherry) [inline redislock_test.go:348]: Nit: maybe would be nicer to check for `errors.Is(err, ErrNotObtained)`, but not critical at all.

**Review by mxmCherry — COMMENTED** [review]
Tests look good.

I think we can merge this and polish (address my nit-comments) ourselves and then tag a next version.

File: msetnxex.lua
Comment (bhallionOhbibi) [reply msetnxex.lua:6]: True, its more straight forward to return true/false or success/error statement instead of a count.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:128]: That would be a neat optimization

**Comment by bhallionOhbibi** [comment]
Thanks for the review and for considering merging this. 
I suppose no changes are needed on my side since you plan on polish this to your taste (which I agree with btw).
I wont touch the code anymore, unless you need me to.

**Comment by mxmCherry** [comment]
This one is now up to @dim (ping-ping). I can take over then, unless dim pushes for Pipelines (IMO, I personally prefer PR-ed Lua approach).

**Comment by bhallionOhbibi** [comment]
Ping

File: msetnxex.lua
Comment (dim) [reply msetnxex.lua:6]: I'm indifferent, I would probably just:

```lua
if redis.call("msetnx", unpack(KEYS)) == 0 then
  return 0
end

for _, key in ipairs(KEYS) do
  # ...
end
return 1
```

File: msetnxex.lua
Comment (dim) [inline msetnxex.lua:4]: Right, this is not right. You should not mix values with keys. Your `KEYS` should only contains they keys. The `value` should be `ARGV[1]` and your `ttl` then `ARGV[2]`

File: pttl.lua
Comment (dim) [reply pttl.lua:5]: I agree, we should return `min` here, but what if we have two keys and only one of the values is not matching?

File: pttl.lua
Comment (dim) [reply pttl.lua:8]: Please fix before merging

File: msetnxex.lua
Comment (dim) [reply msetnxex.lua:6]: Oh, this script needs some updating after the most recent PR #71

File: redislock.go
Comment (dim) [reply redislock.go:70]: Please fix comment

File: redislock.go
Comment (dim) [reply redislock.go:97]: agree, we should not impose any limitations

File: redislock.go
Comment (dim) [inline redislock.go:130]: Sorry, you cannot do that, you cannot pass `value` as redis keys. You need to do this:

```go
res, err := luaMSetNXEX.Run(ctx, c.client, keys, value, ttl.Milliseconds()).Result()
```

File: redislock.go
Comment (dim) [reply redislock.go:173]: That's a breaking change and most people will not use it. Two options:

1. Please keep the original `Key` method and return the first value.
2. Add a `MultiLock` type

File: msetnxex.lua
Comment (dim) [inline msetnxex.lua:2]: This is not right, please see below. you need to construct a new array using `KEYS` with `ARGV[1]` instead

**Review by dim — CHANGES_REQUESTED** [review]
There are quite a few issues that need to be addressed. Please also take a look at #71 as there has been a change that will affect your PR.

Apologies that it took me so long to review this, but it's been a few busy weeks

**Comment by tomaswarynyca** [comment]
@bhallionOhbibi do you have enough time to fix the problems?

File: msetnxex.lua
Comment (bhallionOhbibi) [reply msetnxex.lua:4]: Ah ! yes it is kinda strange, but it keep the script pretty simple and it feels more optimized that way (not having to reconstruct a new array with KEYS + ARGS[1] content).
One comment to explain it better would to the trick perhaps ? Else I'l try to find a way to merge two arrays together...

Comment (dim) [reply msetnxex.lua:4]: sorry, I appreciate the neatness of the code, but keys must be passed as keys to maintain long-term compatibility with redis and be compatible with cluster. if you pass a key as an argument, a redis cluster will not know where to find it.

File: pttl.lua
Comment (bhallionOhbibi) [reply pttl.lua:5]: MIN ?

**Comment by tomaswarynyca** [comment]
I am very pleased to hear that you will be able to continue with the work! :+1: 
@dim @mxmCherry I'm sure they will help you with the questions you have!

File: pttl.lua
Comment (bhallionOhbibi) [reply pttl.lua:5]: @dim I suggest to return the shortest TTL and remove the value check to allow anyone to read this value.

File: pttl.lua
Comment (bhallionOhbibi) [reply pttl.lua:9]: Fixed

File: redislock.go
Comment (bhallionOhbibi) [inline redislock.go:191]: This handling error is more explicit that `redis.Nil` but also not really necessary. Please let me know what you think about it.

File: redislock_test.go
Comment (bhallionOhbibi) [reply redislock_test.go:363]: removed

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go]: Fixed

File: msetnxex.lua
Comment (bhallionOhbibi) [reply msetnxex.lua]: I'v made necessary changes to the script. Please checkout obtain.lua (I renamed the script)

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:171]: Kept it, use it, like it. But we can remove it if you want, ultimately you decide.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:167]: Ooops it look like this comment about the `Key` method is lost.
You made the comment on this code (which replaced an existing `Key` method.
```golang
// Key returns the redis key used by the lock.
func (l *Lock) Keys() []string {
	return l.keys
}
```
I updated the code like this in my last changes:
```golang
// Key returns the redis key used by the lock.
// If the lock hold multiple key, only the first is returned.
func (l *Lock) Key() string {
	return l.keys[0]
}

// Keys returns the redis keys used by the lock.
func (l *Lock) Keys() []string {
	return l.keys
}
```
Assuming users that use many locks will use Keys. And single lock users can use Key. No more breaking changes.

File: pttl.lua
Comment (bhallionOhbibi) [reply pttl.lua]: Also note that in the `func (l *Lock) TTL` method return a ttl of 0 if the `ErrLockNotHeld` error returned by the lua script.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go]: updated error handling strategy. Please let me know what you think.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go]: Same as above.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go]: Same as above.

File: redislock_test.go
Comment (bhallionOhbibi) [reply redislock_test.go]: Removed

File: release.lua
Comment (bhallionOhbibi) [reply release.lua]: Same as above.

File: refresh.lua
Comment (bhallionOhbibi) [reply refresh.lua]: Same as above.

File: redislock_test.go
Comment (bhallionOhbibi) [reply redislock_test.go:369]: updated !

File: msetnxex.lua
Comment (bhallionOhbibi) [reply msetnxex.lua]: You are absolutly right, I updated the code accordingly.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go]: Updated accordingly.

File: msetnxex.lua
Comment (bhallionOhbibi) [reply msetnxex.lua]: Updated !

**Comment by bhallionOhbibi** [comment]
I made a major update to this whole PR.
I think there is no breaking changes and the changes are up to date with the latest version of `main`.

This is my second dive into lua scripting, bear with me if this is not very idiomatic. Feel free to comment on anything of course.
I hope this only needs few changes to be of your liking before merging.

File: pttl.lua
Comment (dim) [inline pttl.lua:13]: this should start as `-1`

File: redislock.go
Comment (dim) [inline redislock.go:125]: Sorry, two things:

1. Comparing errors by text is definitely a "no" :), please use `errors.Is`.
2. I don't get it, how can `luaObtain.Run` ever produce an `ErrNotObtained` in the first place?

File: redislock.go
Comment (dim) [inline redislock.go:195]: same as above, this makes no sense

File: redislock.go
Comment (dim) [inline redislock.go:215]: and again ...

File: redislock.go
Comment (dim) [inline redislock.go:206]: You need to update the comment to explain what is happening when multiple locks are held. What if you hold two locks and one can be extended and the other isn't? This is not a trivial problem.

File: refresh.lua
Comment (dim) [inline refresh.lua:8]: that's NOT a good way to deal with errors

File: obtain.lua
Comment (dim) [inline obtain.lua:18]: sorry, what ma I missing here? why every 2nd key?

File: refresh.lua
Comment (dim) [inline refresh.lua:14]: After reviewing this I actually becoming a little more hesitant to merge this. The problem is that it's actually not trivial to manage multiple locks. Assuming you hold a lock on 3 keys and you are refreshing, the first one succeeds, the second one fails, the third one is never attempted. This behaviour will not be particularly clear to a user. Shouldn't we abandon the first lock in this case too?

Comment (dim) [reply refresh.lua:14]: I am starting to wonder if it's better to allow operating multiple `Lock` instances through a Redis pipeline. This way, each `Lock` holds its own unique state and the user can decide what to do on failures.

**Review by dim — CHANGES_REQUESTED** [review]

**Comment by dim** [comment]
Alternatively, I would not be opposed to having an `ObtainMany` which (atomically) obtains a slice of `[]*Lock`. For example:

```go
type MultiLock []*Lock

func (l MultiLock) MinTTL(ctx context.Context) (time.Duration, error)
func (l MultiLock) Release(ctx context.Context) error

func (c *Client) ObtainMany(ctx context.Context, keys []string, ttl time.Duration, opt *Options) (MultiLock, error)
```

File: refresh.lua
Comment (tomaswarynyca) [reply refresh.lua:14]: It seems to me that if something fails in the process that makes that some of the keys that the user wants to lock is not done successfully, everything has to fail. The lock process with many keys must be easy for the user (while maintaining security).

Comment (dim) [reply refresh.lua:14]: Agree, it should be either all or nothing which is not exactly easy, given that you cannot just revert to the previous state mid-script.

Comment (tomaswarynyca) [reply refresh.lua:14]: Do you think that implementing `MultiLock` will make it easier to roll back in case of failure?

File: obtain.lua
Comment (bhallionOhbibi) [reply obtain.lua:18]: Oups, that's a silly mistake. This line was kept after I removed values from KEYS array.
I'll remove it.

File: pttl.lua
Comment (bhallionOhbibi) [reply pttl.lua:13]: Please note that pttl will return negative value in case of error (like if the key does not exists). Therefore minTTL might never be set. Setting it to zero avoid checking for potential -1 before returning it.
Is there any particular reason to set it to -1 (and therefore add another if statement) ? 
I am missing something ?

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:125]: 1. Cannot use `errors.Is` since error pointer are not the same.
2. It can't, that's why text compare is required.

I'll update this to simply return false and check for `redis: nil` :)

At some point (during development) it made sense for the script to return distinct errors. But it can only result in a `ErrNotObtained` anyway now.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:195]: Updated !

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:215]: Updated !

File: refresh.lua
Comment (bhallionOhbibi) [reply refresh.lua:8]: Updated !

File: refresh.lua
Comment (bhallionOhbibi) [reply refresh.lua:14]: Can the pexpire fail ? it seems to me that the command cannot fail in the key exists: https://redis.io/commands/pexpire/
Assuming the user must owns all the lock (first loop of this script), then pexpire cannot half-fail.

Comment (bhallionOhbibi) [reply refresh.lua:14]: I updated all scripts using pexpire and dont check for pexpire error anymore.
What do you think ?
Of course a server or connexion issue can always be involved, but I'm not sure its possible while executing the lua script on the server.

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:206]: How would one lock be extended and not the other ?
Given the atomicity of the script, either both lock exists and you can refresh them or at least one is missing and none are refreshed.

**Comment by bhallionOhbibi** [comment]
About multi-lock and hypothetical partially held locks:

Theorically, you cannot hold one lock and not the other. `msetnx` make sure that you either set all the locks or none.
Also, all locks held together share the same value and ttl.
Therefore once you know that you and only you own all the locks, any of the commands (Refresh, TTL, Release) will come together nicely.
- Refresh: you own them all, you refresh them all.
- TTL: they all have the same ttl (minTTL check is really just for the show)
- Release: Same as refresh.

And for each of theses, if somehow you dont hold one lock, an error is returned and no refresh or release is done on any of the key among the multilock.

**Comment by bhallionOhbibi** [comment]
About multi-lock and hypothetical partially held locks:

Theorically, you cannot hold one lock and not the other. `msetnx` make sure that you either set all the locks or none.
Also, all locks held together share the same value and ttl.
Therefore once you know that you and only you own all the locks, any of the commands (Refresh, TTL, Release) will come together nicely.
- Refresh: you own them all, you refresh them all.
- TTL: they all have the same ttl (minTTL check is really just for the show)
- Release: Same as refresh.

And for each of theses, if somehow you dont hold one lock, an error is returned and no refresh or release is done on any of the key among the multilock.

**Comment by tomaswarynyca** [comment]
@dim ping, what do you think of the changes made?

**Comment by dim** [comment]
I will take a look next week. I think most of my concerns have been addressed through the new `release.lua`, but need to double-check.

File: obtain.lua
Comment (dim) [inline obtain.lua:17]: minor: `i` not used, should be `_`

File: obtain.lua
Comment (dim) [inline obtain.lua:27]: same as above

File: redislock.go
Comment (dim) [inline redislock.go:68]: minor: don't you think `ObtainMulti` sounds better? also, more consistent with go stdlib

File: redislock.go
Comment (dim) [inline redislock.go:188]: we should mention that it returns the minimum TTL if multiple keys are held by the lock

File: release.lua
Comment (dim) [inline release.lua:15]: this is a Go, not a Redis type error. you should wrap in Go code

**Review by dim — CHANGES_REQUESTED** [review]
Almost there, I am OK with the changes now, just a few minor issues left. Could you please update the changelog too?

**Comment by dim** [comment]
It's a relatively big change, I think `v0.10.0` is justified

**Comment by tomaswarynyca** [comment]
ping @bhallionOhbibi

**Comment by bhallionOhbibi** [comment]
Sorry ! I was on a very long vacation during october. I'll check your feedback soon. Thanks

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go:188]: Done !

File: obtain.lua
Comment (bhallionOhbibi) [reply obtain.lua]: Fixed

File: obtain.lua
Comment (bhallionOhbibi) [reply obtain.lua]: Fixed

File: redislock.go
Comment (bhallionOhbibi) [reply redislock.go]: Renamed

File: release.lua
Comment (bhallionOhbibi) [reply release.lua]: I simplified this part. We dont really need to check error for `del` and the loop above already ensure the key exists.

**Comment by bhallionOhbibi** [comment]
Ping :)

**Review by dim — APPROVED** [review]

**Comment by bhallionOhbibi** [comment]
Thank you for all your feedbacks and time dedicated to this

**Comment by dzvancuks** [comment]
Hello. When will the next tag be released? Currently it is still [v0.9.4](https://github.com/bsm/redislock/releases/tag/v0.9.4) that don't have these changes.

**Pull request merged**
